source/arm11/drivers/pdn.c

```diff
@@ -212,4 +213,25 @@ void PDN_controlGpu(const bool enableClk, const bool resetPsc, const bool resetO
 		wait_cycles(12); // 4x the needed cycles in case we are in LGR2 mode.
 		pdn->gpu_cnt = reg | PDN_GPU_CNT_NORST_ALL;
 	}
+}
+
+void PDN_sleep(void)
+{
+	getPdnRegs()->wake_enable = PDN_WAKE_SHELL_OPENED;
+	getPdnRegs()->cnt |= PDN_CNT_SLEEP;
+}
+
+void PDN_wakeup(void)
+{
+	static KHandle pdnWakeEvent;
+	if (!pdnWakeEvent)
+	{
+		pdnWakeEvent = createEvent(true);
+	}
+
+	bindInterruptToEvent(pdnWakeEvent, IRQ_PDN, 14);
+	waitForEvent(pdnWakeEvent);
+	getPdnRegs()->wake_enable = 0;
+	getPdnRegs()->wake_reason = PDN_WAKE_SHELL_OPENED;
+	unbindInterruptEvent(IRQ_PDN);
```

Comment:
Let me summarize what i know about hardware sleep mode:

HOS first notifies the sysmodules about sleep so they can prepare sleep mode. This is for example the GFX_sleep() function that's being called. Other ones to call are CODEC_deinit() (poorly documented, sleep is basically the same as deinit). And the power LED is set to the sleep mode animation via the MCU.
After all the hardware we don't need in sleep mode is in sleep mode Kernel11 does check this bit:
https://github.com/profi200/libn3ds/blob/master/include/arm11/drivers/pdn.h#L71

If it's set it does extra steps to disable VRAM banks and clock via:
https://github.com/profi200/libn3ds/blob/master/include/arm11/drivers/pdn.h#L103

Then it disables FCRAM clock. In GBA mode we already did this but just for reference:
https://github.com/profi200/libn3ds/blob/master/include/arm11/drivers/pdn.h#L114
https://github.com/profi200/libn3ds/blob/master/source/arm11/drivers/lgy11.c#L59

Then finally comes the sleep mode part. This bit is set and the CPU has to enter wait for interrupt state somewhere. This would likely be the kernel idle thread if nothing has to be done:
https://github.com/profi200/libn3ds/blob/master/include/arm11/drivers/pdn.h#L70

I think PDN swake reasons are managed externall by some sysmodule. Not sure.

Reply:
I have modified this part again.

I have moved the PDN related function calls from gfx.c to pdn.c, so disabling GPU clock and VRAM now is handled by pdn.c

I haven't implemented disabling FCRAM clock mostly because I'm not familiar with it, and my motivation is mostly on the GBA side, but please let me know if here's any reference that I can follow, especially for re-enabling it after waking up.

After testing with the sleep mode again, it seems that calling _wfi is enough to ensure that it stays asleep until next interrupt, which is when the lid is opened again. I haven't found other source of interrupt other than opening the lid when it's sleeping, so I'm guessing it would work.
As for PDN wake reasons, from my testing, it doesn't seem to do anything other than flags for the developer. i.e., I could change it to `PDN_WAKE_TOUCHPEN_DOWN` and it will still behave the same.

I have reverted the previous use of bindEvent on IRQ_PDN, since it seems there are different expectations of what should happen during the ISR of sleep mode.